### PR TITLE
add SCGEN to models & plotting.py to plotting

### PR DESCRIPTION
**Pull request by aidinbii**
I added scgen to models folder. Also plotting.py to plotting folder.

### CODE REVIEW

**Comment by M0hammadL** [comment]
@aidinbii  can you please update examples: rerun this example with scgen:
https://github.com/theislab/scarches/blob/master/notebooks/scanvi_surgery_pipeline.ipynb

same dataset as query and reference

**Comment by M0hammadL** [comment]
@aidinbii I tested your notebook as wrote in our chat, please do not change that and add redundant stuff, please correct this : 

- [ ] There is no option of returning corrected latent space in ` corrected_reference_adata = network.batch_removal(source_adata, batch_key="study", cell_label_key="cell_type",return_latent=True)` and it doens not work, as I wrote please put it in `adata.obsm["corrected"]` also original save gene expression in `adata.layers["original_data"]`. these features need to be also in  ` sca.models.SCGEN.map_query_data ` add  `return_latent` also in here. 

- [ ] make sure corrected latent space is also working  (plot the results in the example) and agree with the corrected gene expression

- [ ] please change `sca.models.SCGEN` to `sca.models.scgen`
- [ ] make sure when you store original uncorrected data in `adata.layers["original_data"]` the orders of the cell do not get shuffled meaning that you have same order in adata.X as in adata

- [ ]  test if the ` sca.models.SCGEN.map_query_data ` works if you also pass model object instead of reading from file 
- [ ] please show all these tests in separate notebook other than I shared, the on I shared just reproduce and add show corrected latent pace in addition to corrected gene expression.
- [ ] remove plotting, we don't need it for scarches

**Comment by M0hammadL** [comment]
@aidinbii  looks good a few things : 

- [ ] corrected_reference_adata.obsm["corrected"] call it `corrected_reference_adata.obsm["latent_corrected"]  and mention it is saved there
- [ ] put a raw uncorrected data in adata.obs["original_data"]  and mention in the notebook it is there
- [ ]  you don't need to create another adata for latent : 

```
corrected_reference_latent = sc.AnnData(corrected_reference_adata.obsm["corrected"])
corrected_reference_latent.obs["cell_type"] = source_adata.obs[cell_type_key].tolist()
corrected_reference_latent.obs["batch"] = source_adata.obs[condition_key].tolist()
sc.pp.neighbors(corrected_reference_latent)
sc.tl.umap(corrected_reference_latent)
sc.pl.umap(corrected_reference_latent,
           color=['batch', 'cell_type'],
           frameon=False,
           wspa
ce=0.6,
           )```
instead just do 

```
sc.pp.neighbors(corrected_reference_adata, use_rep="latent_corrected")
sc.tl.umap(corrected_reference_latent)
sc.pl.umap(corrected_reference_latent,
           color=['batch', 'cell_type'],
           frameon=False,
           wspa
ce=0.6,
           )
```

- [ ] also show latent representation of  query and reference right now you only show it for reference data, add it and the end

File: scarches/models/scgen/vaearith_model.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+        init_params = {
+            'hidden_layer_sizes': dct['hidden_layer_sizes_'],
+            'z_dimension': dct['z_dim_'],
+            'dr_rate': dct['dr_rate_'],
+        }
+
+        return init_params
+
+    @classmethod
+    def load_query_data(cls,
```
Comment (Koncopd) [inline scarches/models/scgen/vaearith_model.py:324]: Again not needed here, what to expand and freeze if no conditional weights are present.

File: scarches/models/scgen/vaearith_model.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+        self.is_trained_ = False
+        self.trainer = None
+
+    def train(self, n_epochs = 25, batch_size = 32, **kwargs):
+        self.trainer = vaeArithTrainer(self.model, self.adata, n_epochs, batch_size, **kwargs)
+        self.trainer.train()
+        self.is_trained_ = True
+
+
+    def to_latent(self, data: Optional[np.ndarray] = None):
```
Comment (Koncopd) [inline scarches/models/scgen/vaearith_model.py:205]: rename to get_latent

File: scarches/models/scgen/vaearith_model.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        model = cls(adata, **init_params)
+        model.model.load_state_dict(model_state_dict)
+        model.model.eval()
+
+        model.is_trained_ = attr_dict['is_trained_']
+
+        return model
+
+
+class scgen(BaseMixin):
```
[the remark this answers is not in the document]
Comment (M0hammadL) [reply scarches/models/scgen/vaearith_model.py:168]: > should be scGen i think. @M0hammadL ?

scgen also fine

**Comment by M0hammadL** [comment]
@aidinbii  did you apply @Koncopd  code corrections above?

File: scarches/models/scgen/vaearith_model.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+
+        torch.save(self.model.state_dict(), model_save_path)
+
+    def _load_expand_params_from_dict(self, state_dict):
+        load_state_dict = state_dict.copy()
+
+        device = next(self.model.parameters()).device
+
+        new_state_dict = self.model.state_dict()
+        for key, load_ten in load_state_dict.items():
```
Comment (Koncopd) [inline scarches/models/scgen/vaearith_model.py:94]: still there is no need to expand anything, just load_state_dict should suffice.

File: scarches/models/scgen/trainer.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
+import torch
```
Comment (Koncopd) [inline scarches/models/scgen/trainer.py:1]: Could you move this file to trainers/scgen

**Comment by M0hammadL** [comment]
have you also checked it also works if you pass the module object to the map query instead of reading from the file.

**Comment by aidinbii** [comment]
> have you also checked it also works if you pass the module object to the map query instead of reading from the file.
Yes, I checked it in the test notebook.

**Comment by aidinbii** [comment]
> @aidinbii did you apply @Koncopd code corrections above?



> > have you also checked it also works if you pass the module object to the map query instead of reading from the file.
> 
> Yes, I checked it in the test notebook.

**Comment by M0hammadL** [comment]
@Koncopd  are we good to go? Can you do the final check, if you give the signal then I merge.

@aidinbii please reproduce the notebook again since you moved around files and etc to make sure it works again

**Comment by aidinbii** [comment]
@M0hammadL The notebooks are working

**Comment by M0hammadL** [comment]
@Koncopd  good to go? can you do a final check?

File: scarches/models/scgen/vaearith_model.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+        attr_save_path = os.path.join(dir_path, "attr.pkl")
+        varnames_save_path = os.path.join(dir_path, "var_names.csv")
+
+        var_names = self.adata.var_names.astype(str)
+        var_names = var_names.to_numpy()
+        np.savetxt(varnames_save_path, var_names, fmt="%s")
+
+        torch.save(self.model.state_dict(), model_save_path)
+
+    def _load_expand_params_from_dict(self, state_dict):
```
Comment (Koncopd) [inline scarches/models/scgen/vaearith_model.py:88]: no point in having it at all, just delete it and replace with direct call of `load_state_dict`.

File: scarches/models/scgen/vaearith_model.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        Returns
+        -------
+        integrated: `~anndata.AnnData`
+        Returns an integrated query.
+        """
+        if isinstance(reference_model, str):
+            attr_dict, model_state_dict, var_names = cls._load_params(reference_model)
+            _validate_var_names(query, var_names)
+        else:
+            attr_dict = reference_model._get_public_attributes()
```
Comment (Koncopd) [inline scarches/models/scgen/vaearith_model.py:335]: There is no point in reinitialization of the model, we don't expand anything. Just use the model if it is passed:
if isinstance(reference_model, str):
  load the model as new_model
else:
  new_model = reference_model

**Comment by Koncopd** [comment]
@M0hammadL @aidinbii added some comments.

**Comment by aidinbii** [comment]
@Koncopd added corrections and checked that they are working.

**Comment by M0hammadL** [comment]
> @M0hammadL @aidinbii added some comments.

how about now?

**Comment by Koncopd** [comment]
@M0hammadL @aidinbii ok, seems ready now, can be merged if everything works correctly.
